### DISCUSSION TITLE
TPWR for MPM

### DISCUSSION THREAD

**Discussion by inventor7777** [EdgeTX/edgetx#5131] [Q&A] [answered]
Hi all! I have seen that ExpressLRS has a telemetry sensor built in for showing Tx power, called TPWR, but I don’t see it on my MPM FCC RadioMaster MT12. Is there any way to get it on my radio? I asked under the MPM repo but I haven’t gotten any answer, so I’ll ask here. If you need any files from radio let me know. 

Here is the discussion on the MPM repo. https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/discussions/984

**Comment by pfeerick** [comment, marked as answer]
We only show sensors the MPM exposes, so not until it supports that. I'm no
ot sure it's appropriate or needed, since the MPM only has two power
levels, full and low. It doesn't have multiple levels like CRSF/ELRS, etc.

**Reply by inventor7777** [reply]
Ok, makes sense. Based off of this code, I at first assumed it had multiple levels, but after reexamination I suppose you’re probably right. I still want to know when it switches from low to high, but I think it’s based off of signal DBm, so I assume I could calculate it based off of that when I have more time to read the code.
![IMG_3340](https://github.com/EdgeTX/edgetx/assets/142270525/e0a5cf14-cb74-4a83-b353-e0930707824e)

**Comment by pfeerick** [comment]
But did you check where/if that code is actually used?

**Reply by inventor7777** [reply]
Pascal sent it to me in a previous discussion, so I’m assuming yes.
![IMG_3343](https://github.com/EdgeTX/edgetx/assets/142270525/18f75bf7-4355-47f2-b9a0-21f65d30ded9)

**Reply by pfeerick** [reply]
That would be a "no", then. If you actually searched for say... "TXPOWER_300uW", you'd see it is commented out code ;) 

You need  to scroll further down, since you are actually interested in the lines below for A7105_POWER, NRF_POWER, CC2500_POWER, CYRF_POWER, There is only is only high power and low power, in addition to internal use range check and bind power settings. ;)

**Reply by inventor7777** [reply]
I thought that’s what I said? 
>I at first assumed it had multiple levels, but after reexamination I suppose you’re probably right

But am I right in assuming it chooses its power based off of dBm? Or do I need more sleep 😆

**Reply by pfeerick** [reply]
you also said
> Pascal sent it to me in a previous discussion, so I’m assuming yes.

but I didn't see the underline before on the phone indicating you *were* looking at the correct bit of code... sorry! 

> But am I right in assuming it chooses its power based off of dBm? Or do I need more sleep 😆

Yes, and no. It doesn't "choose it", but the power level is *set* to a power level from the range each RC transceiver supports. HIGH POWER is set to a specific value for each RF transceiver chip, based off it's capabilities, as is LOW POWER, etc, etc. This is _before_ it goes to any RF amplifier end stage, so does not mean you'll get that output at the antenna. i.e. for the NRF24, it only has four steps in it's internal PA... whereas some others like the CC2500 have seventeen.

![image](https://github.com/EdgeTX/edgetx/assets/5500713/84559632-5b13-4efb-942b-31a2ba1f38d2)

**Reply by inventor7777** [reply]
Mmm, ok, makes sense. Sorry if I wasn’t clear. And thanks for the PA info.

But it _does_ choose between low and high power, correct? Even if there isn’t an absolute, concrete output power value it does still have two power modes (not counting bind and range test mode), right?

**Reply by pfeerick** [reply]
Yes, and that is what the **Low Power** option I told you about before in the MPM options does ;)

**Reply by inventor7777** [reply]
Ok, thank you very much for your time!

**Reply by pfeerick** [reply]
No problem. Don't stop asking questions... this is a big and sometimes complex hobby ;)
